File: Programmers/level1/dartGame/dart_game.py
# 문제 출처 : https://programmers.co.kr/learn/courses/30/lessons/17682

import re


def solution(dartResult):
    answer = 0
    scores = []
    num = ''
    p = re.compile('(\d+)([SDT])([*#]?)')
    dart = p.findall(dartResult)
    print(dart)
    for dart in dartResult:
        if dart == "S":
            scores.append(int(num))
            scores[-1] = scores[-1] ** 1
            num = ''
        elif dart == "D":
            scores.append(int(num))
            scores[-1] = scores[-1] ** 2
            num = ''
        elif dart == "T":
            scores.append(int(num))
            scores[-1] = scores[-1] ** 3
            num = ''
        elif dart == "*":
            if len(scores) > 1:
                scores[-1] = scores[-1] * 2
                scores[-2] = scores[-2] * 2
            else:
                scores[-1] = scores[-1] * 2
        elif dart == "#":
            scores[-1] = scores[-1] * -1
        else:
            num += dart
    print(scores)
    return sum(scores)


solution("1S2D*3T")

'''
다른 코드 : https://programmers.co.kr/learn/courses/30/lessons/17682/solution_groups?language=python3

import re

def solution(dartResult):
    bonus = {'S' : 1, 'D' : 2, 'T' : 3}
    option = {'' : 1, '*' : 2, '#' : -1}
    p = re.compile('(\d+)([SDT])([*#]?)')
    dart = p.findall(dartResult)
    for i in range(len(dart)):
        if dart[i][2] == '*' and i > 0:
            dart[i-1] *= 2
        dart[i] = int(dart[i][0]) ** bonus[dart[i][1]] * option[dart[i][2]]

    answer = sum(dart)
    return answer
'''

File: Programmers/level1/dartGame/dartGame.js
// https://programmers.co.kr/learn/courses/30/lessons/17682?language=javascript

export const solution = (dartResult) => {
    let bonus = { "S": 1, "D": 2, "T": 3};
    let options = { "*": 2, "#": -1, null: 1};
    let re = dartResult.match(/\d.?\D/g);
    let scoreArr = [];
    // console.log(re)
    re.forEach(el => {
        let num = Number(el.match(/[\d]+/g)[0]);
        let bon = el.match(/[SDT]/g)[0];
        let op = el.match(/[*#]/g);
        // console.log(`숫자 : ${num}, 보너스 : ${bon}, 옵션 : ${op}`)
        let score = Math.pow(num, bonus[bon]) * options[op];
        
        if(options[op] === 2){
            if(scoreArr.length === 0){
                scoreArr.push(score);
            } else {
                let pop = scoreArr.pop();
                scoreArr.push(pop * 2);
                scoreArr.push(score);
            }
        } else {
            scoreArr.push(score);
        }
    })
    // console.log(scoreArr)
    return scoreArr.reduce((a,b) => a + b);
}